x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/InternalMapReduceAggregation.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.mapreduce;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+import org.elasticsearch.search.aggregations.AggregationReduceContext;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.support.SamplingContext;
+import org.elasticsearch.search.profile.SearchProfileResults;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.ToXContent.DelegatingMapParams;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+public final class InternalMapReduceAggregation<
+    MapContext extends Closeable,
+    MapFinalContext extends Writeable,
+    ReduceContext extends Closeable,
+    Result extends ToXContent & Writeable> extends InternalAggregation {
+
+    private final AbstractMapReducer<MapContext, MapFinalContext, ReduceContext, Result> mapReducer;
+    private final BigArrays bigArraysForMapReduce;
+    private final List<String> fieldNames;
+    private final boolean profiling;
+
+    private MapFinalContext mapFinalContext = null;
+    private Result mapReduceResult = null;
+
+    InternalMapReduceAggregation(
+        String name,
+        Map<String, Object> metadata,
+        AbstractMapReducer<MapContext, MapFinalContext, ReduceContext, Result> mapReducer,
+        MapFinalContext mapFinalContext,
+        Result mapReduceResult,
+        List<String> fieldNames,
+        boolean profiling
+    ) {
+        super(name, metadata);
+        this.mapReducer = Objects.requireNonNull(mapReducer);
+        this.mapFinalContext = mapFinalContext;
+        this.mapReduceResult = mapReduceResult;
+        this.fieldNames = fieldNames;
+        this.profiling = profiling;
+
+        // we use the `NON_RECYCLING_INSTANCE` here, which has no circuit breaker attached
+        // However circuit breaking works different in this place:
+        // {@link QueryPhaseResultConsumer} reserves the serialized size of this object plus 50% extra
+        // using the non-breaking and non-recycling instance is ok under the assumption that the real memory
+        // usage is not more than 1.5 * the serialized size
```

Comment:
I am pretty uncomfortable with that assumption here.  I get that you've tested that, but that's a far cry from knowing for sure that it's reliable under all user work loads.